### ISSUE TITLE
Seaside-Core, Seaside-Session and Seaside-Environment have gemstone-specific branches

### ISSUE THREAD

**Issue by jbrichau** [SeasideSt/Seaside#845]
It would be better to get rid of the branched packages to ease the porting to Gemstone. 

While I am now creating a gemstone branch on the SeasideSt/Seaside repo to port 3.2.0 to Gemstone, I'm giving these differences a close look and try to adapt the codebase to remove them. 

@dalehenrich fyi, I'm pulling in the changes for 3.2.0 in the current 3.1.x gs-master codebase and will be using the gemstone branch in this repo as a temporary place.

**Comment by jbrichau** [comment]
Differences between the .gemstone package branch and the main branch:

### Seaside-Session
`WAMutualExclusionFilter>>initialize` has a WAGemstoneMutex class ref

### Seaside-Core
`WAGemStoneRequestContext` class being used 
`WAServerManager class>>default` uses session temps
`WATableBasedEncoder>>nextPutAll:` has a GS-specific optimization

**Comment by jbrichau** [comment]
So, `WAGemStoneMutex` will be removed because @marschall is going to implement WAMutex with a `TransientValue` holder for the owner instvar. 

The rest is TBD tomorrow :)

**Comment by dalehenrich** [comment]
diffs for Seaside-Core-pmm.818 vs Seaside-Core.gemstone-JohanBrichau.818 (latest versions in https://github.com/GsDevKit/Seaside31):

```diff

Item was changed:
  ----- Method: WACache>>initialize (in category 'initialization') -----initialize
    super initialize.
    self initializeCollections.
    self initializeMutex.
+   self setExpiryPolicy: WALastAccessExpiryPolicy new.
+   self setReapingStrategy: WAAccessIntervalReapingStrategy new.
-   self setExpiryPolicy: WARcLastAccessExpiryPolicy new.
-   self setReapingStrategy: WANoReapingStrategy new.
    self setMissStrategy: WACacheMissStrategy new.
    self setRemovalAction: WANotifyRemovalAction new!

Item was changed:
  ----- Method: WACache>>initializeMutex (in category 'initialization') -----initializeMutex
+   mutex := WAMutex new!
-   mutex := WAGemStoneMutex new!

Item was changed:
  ----- Method: WAServerAdaptor>>contextFor: (in category 'converting') -----contextFor: aNativeRequest
    "Answer a request context for aNativeRequest."
  
+   ^ WARequestContext
-   ^ WAGemStoneRequestContext
      request: (self requestFor: aNativeRequest)
      response: (self responseFor: aNativeRequest)
      codec: self codec!

Item was changed:
  ----- Method: WAServerManager classSide>>default (in category 'accessing') -----default
+   ^ default ifNil: [ default := self new ]!
-   | singleton |
-   singleton := SessionTemps current
-     at: #'WA_SERVER_MANAGER_DEFAULT'
-     otherwise: nil.
-   singleton == nil
-     ifTrue: [ 
-       singleton := self new.
-       SessionTemps current at: #'WA_SERVER_MANAGER_DEFAULT' put: singleton ].
-   ^ singleton!

Item was removed:
- WASystemConfiguration subclass: #WAStrictTransportSecurityFilterConfiguration	instanceVariableNames: ''	classVariableNames: ''	poolDictionaries: ''	category: 'Seaside-Core-Filter'!!WAStrictTransportSecurityFilterConfiguration commentStamp: 'pmm 2/9/2014 12:38' prior: 0!I am the configuration for WAStrictTransportSecurityFilter.!

Item was removed:
- ----- Method: WAStrictTransportSecurityFilterConfiguration>>describeOn: (in category 'description') -----describeOn: config
-   (config integer: #'maxAge')
-     comment:
-         'The number of seconds the user agent should remember receipt of this header field from this server.';
-     default: 3600 * 24.	"one day"
-   (config boolean: #'includeSubDomains')
-     comment:
-         'If true signals to the user agent that the STS Policy applies to this STS Server as well as any subdomains of the server''s fully qualified domain name.';
-     default: false!

Item was changed:
  ----- Method: WATableBasedEncoder>>nextPutAll: (in category 'accessing') -----nextPutAll: aString
+   "uses #to:do: for speed reasons (on Pharo)
+ 	this is not premature optimization, this is a hotspot method method
+ 	and #to:do: shows measurable speed improvements for rendering seaside pages"
-   "#encodeUsing: is a primitive in GemStone, so it does the mapping very quickly..."
  
+   1 to: aString size do: [ :index | self nextPut: (aString at: index) ]!
-   | encodedString |
-   encodedString := aString encodeUsing: table.
-   encodedString ~~ nil
-     ifTrue: [ stream nextPutAll: encodedString ]
-     ifFalse: [ 1 to: aString size do: [ :index | self nextPut: (aString at: index) ] ]!
```

**Comment by dalehenrich** [comment]
No differences between Seaside-Environment-pmm.81 and Seaside-Environment.gemstone-dkh.81  ... I must have anticipated changes that never meterialized...

**Comment by dalehenrich** [comment]
diffs for Seaside-Core-pmm.818 vs Seaside-Core.gemstone-JohanBrichau.818  (latest versions in https://github.com/GsDevKit/Seaside31):

```diff
Item was changed:
  ----- Method: WACache>>initialize (in category 'initialization') -----initialize
    super initialize.
    self initializeCollections.
    self initializeMutex.
+   self setExpiryPolicy: WALastAccessExpiryPolicy new.
+   self setReapingStrategy: WAAccessIntervalReapingStrategy new.
-   self setExpiryPolicy: WARcLastAccessExpiryPolicy new.
-   self setReapingStrategy: WANoReapingStrategy new.
    self setMissStrategy: WACacheMissStrategy new.
    self setRemovalAction: WANotifyRemovalAction new!

Item was changed:
  ----- Method: WACache>>initializeMutex (in category 'initialization') -----initializeMutex
+   mutex := WAMutex new!
-   mutex := WAGemStoneMutex new!

Item was changed:
  ----- Method: WAServerAdaptor>>contextFor: (in category 'converting') -----contextFor: aNativeRequest
    "Answer a request context for aNativeRequest."
  
+   ^ WARequestContext
-   ^ WAGemStoneRequestContext
      request: (self requestFor: aNativeRequest)
      response: (self responseFor: aNativeRequest)
      codec: self codec!

Item was changed:
  ----- Method: WAServerManager classSide>>default (in category 'accessing') -----default
+   ^ default ifNil: [ default := self new ]!
-   | singleton |
-   singleton := SessionTemps current
-     at: #'WA_SERVER_MANAGER_DEFAULT'
-     otherwise: nil.
-   singleton == nil
-     ifTrue: [ 
-       singleton := self new.
-       SessionTemps current at: #'WA_SERVER_MANAGER_DEFAULT' put: singleton ].
-   ^ singleton!

Item was removed:
- WASystemConfiguration subclass: #WAStrictTransportSecurityFilterConfiguration	instanceVariableNames: ''	classVariableNames: ''	poolDictionaries: ''	category: 'Seaside-Core-Filter'!!WAStrictTransportSecurityFilterConfiguration commentStamp: 'pmm 2/9/2014 12:38' prior: 0!I am the configuration for WAStrictTransportSecurityFilter.!

Item was removed:
- ----- Method: WAStrictTransportSecurityFilterConfiguration>>describeOn: (in category 'description') -----describeOn: config
-   (config integer: #'maxAge')
-     comment:
-         'The number of seconds the user agent should remember receipt of this header field from this server.';
-     default: 3600 * 24.	"one day"
-   (config boolean: #'includeSubDomains')
-     comment:
-         'If true signals to the user agent that the STS Policy applies to this STS Server as well as any subdomains of the server''s fully qualified domain name.';
-     default: false!

Item was changed:
  ----- Method: WATableBasedEncoder>>nextPutAll: (in category 'accessing') -----nextPutAll: aString
+   "uses #to:do: for speed reasons (on Pharo)
+ 	this is not premature optimization, this is a hotspot method method
+ 	and #to:do: shows measurable speed improvements for rendering seaside pages"
-   "#encodeUsing: is a primitive in GemStone, so it does the mapping very quickly..."
  
+   1 to: aString size do: [ :index | self nextPut: (aString at: index) ]!
-   | encodedString |
-   encodedString := aString encodeUsing: table.
-   encodedString ~~ nil
-     ifTrue: [ stream nextPutAll: encodedString ]
-     ifFalse: [ 1 to: aString size do: [ :index | self nextPut: (aString at: index) ] ]!
```

**Comment by dalehenrich** [comment]
diffs forSeaside-Session-pmm.170 vs Seaside-Session.gemstone-JohanBrichau.169:
```diff
Item was changed:
  ----- Method: WAApplication>>handleDefault:withKey: (in category 'handling') -----handleDefault: aRequestContext withKey: aKey
+   "starts a new session with the given key
+ 	Pay attention if the client can choose the key then you are vulnerable to session fixation attacks"
+ 
    self handle: aRequestContext registering: self newSession at: aKey!

Item was changed:
  ----- Method: WACookieSessionTrackingStrategy>>keyFromContext: (in category 'public') -----keyFromContext: aRequestContext
+   "If a query field is present it overrides a cookie."
-   "If a query field is present it overrides a cookie.
- 	In the case of a document handler both a cookie and a query field are present with different values
- 	(the cookie 'pointing' to the session and the query field 'pointing' to the document handler).
- 	We have to take the document handler in this case."
  
    | key |
    key := self queryFieldKeyFromContext: aRequestContext.
    ^ key isNil
      ifTrue: [ (self cookieFromContext: aRequestContext ifAbsent: [ ^ nil ]) value ]
      ifFalse: [ key ]!

Item was changed:
  ----- Method: WAInvisibleSessionTrackingStrategy>>keyFromContext: (in category 'public') -----keyFromContext: aRequestContext
+   "If a query field is present it overrides a session id."
-   "If a query field is present it overrides a session id.
- 	In the case of a document handler both a session id and a query field are present with different values
- 	(the session id 'pointing' to the session and the query field 'pointing' to the document handler).
- 	We have to take the document handler in this case."
  
    | key |
    key := self queryFieldKeyFromContext: aRequestContext.
    ^ key isNil
      ifTrue: [ self sessionIdFromContext: aRequestContext ]
      ifFalse: [ key ]!

Item was changed:
  ----- Method: WAMutualExclusionFilter>>initialize (in category 'initialization') -----initialize
    super initialize.
+   mutex := WAMutex new!
-   mutex := WAGemStoneMutex new!

Item was changed:
  ----- Method: WASession>>createCache (in category 'initialization') -----createCache
    "To configure the continuation cache you must currently subclass WASession and
  	override this method."
  
    ^ WACache new
+     setExpiryPolicy: WALeastRecentlyUsedExpiryPolicy new;
-     setExpiryPolicy: WARcLastAccessExpiryPolicy new;
      yourself!

Item was changed:
  ----- Method: WASession>>initializeFilters (in category 'initialization') -----initializeFilters
-   "For GemStone, mutual exclusion is handled by an object lock, see handle:.
- 	 Over the long haul, I suppose that I could specialize the MutualExclusionFilter for GemStone..."
- 
-   true
-     ifTrue: [ ^ self ].
    self addFilter: WAMutualExclusionFilter new!
```

**Comment by dalehenrich** [comment]
Looks like the changes are pretty straightforward ...

**Comment by jbrichau** [comment]
- Trivial change to remove Seaside-Environment.gemstone done in BaselineOfSeaside3
- Pushed down creation of WAGemStoneRequestContext to GS-specific server adaptors
- WAMutex was adapted in the main code base to work like WAGemStoneMutex
- Expiring and reaping policy no longer required because BulkReaping implementation is now part of main code base (specifically usable for GS)

Remain todo:
- WASession>>initializeFilters
- WATableBasedEncoder>>nextPutAll:
- WAServerManager classSide>>default

**Comment by jbrichau** [comment]
@marschall I wonder if the optimization in WATableBasedEncoder is still necessary? It was when WAXmlEncoder was a subclass, but now only WAUrlEncoder is a subclass. So this GS-specific change could maybe just be removed.

**Comment by jbrichau** [comment]
So, we replaced the direct use of SessionTemps in WAServerManager with a class var holding a TransientValue instance (which eventually uses SessionTemps anyway). That allowed to remove the last difference (I think) between Seaside-Core and Seaside-Core.gemstone.

The use WAMutualExclusionFilter (in Seaside-Session) remains to be done.

**Comment by jbrichau** [comment]
@dalehenrich we're ready with this. So far, it seems everything is working fine. The builds are failing for other reasons (see travis). Will do some more extensive testing in the following days and hopefully you can give the changes a good glance too. 

Once this (and the remaining failing tests) are fixed, we're ready to release 3.2 :)

**Comment by dalehenrich** [comment]
Okay .. I'll take Seaside3.3 for a spin with GemStone 3.2.7 and GemStone 3.3 and skim the changes as well ...

**Comment by dalehenrich** [comment]
woops Seaside3.2 :)

**Comment by dalehenrich** [comment]
Looks like the WAExternalFileLIbraryTests are failing due to a missing: GRPlatform>>defaultDirectoryPathString.

I'm seeing a few more failures (in the gemserver tests) that I'm running locally:

```
3.2.7:
  1322 run, 1313 passes, 0 expected defects, 1 failures, 8 errors, 0 unexpected passes
3.3.0:
  1322 run, 1311 passes, 0 expected defects, 3 failures, 8 errors, 0 unexpected passes
```

The sent but not implemented and undefined symbols are coming from WAGettextExample (both  errors are logged to the transcript):

```
--transcript--'Undefined Symbol:#''WAGetTextExporter'''
--transcript--'Sent but not implemented from #''WAGettextExample class>>export'''
--transcript--'#''exportTemplate'''
```
I will address the the GemServer tests (the other failures) today...

**Comment by jbrichau** [comment]
Hey Dale,

Don't bother about the failing tests. I can fix them. Would be nice if you can check how we refactored the package diffs. See if any issue pops to mind. 

But anything helps and is appreciated of course. So do whatever you feel like :)

Regards from Brescia!

Johan (sent from my mobile)

**Comment by dalehenrich** [comment]
Understood, but I'm curious what's going on with the gem server tests:) ... I'll peek at the package refactoring right now ...

**Comment by dalehenrich** [comment]
Looks like the gemServer failures are random ... which is an issue, but not necessarily a Seaside3.2 issue ...

**Comment by dalehenrich** [comment]
[Made this comment on WAGsUrlEncoder>>nextPutAll:](https://github.com/SeasideSt/Seaside/commit/622616aee2896a8e61ebb0edfa4f622807d6f5af#commitcomment-12213675)

**Comment by dalehenrich** [comment]
Looks like **WAGemStoneMutex** can be deleted or deprecated?

**Comment by dalehenrich** [comment]
I've finished my review and other than the few minor points and Issue #848, things make a lot of sense ... nice to see the gemstone branch done away with! Issue #848 isn't a major problem - it was only exposed when someone tried to reap expired sessions after a year ...

**Comment by marschall** [comment]
Yes `WAGemStoneMutex` is no longer used.

**Comment by jbrichau** [comment]
Addressed comment on WAGsUrlEncoder in 4e0fae7c3e0f0c5afa90f34d8489ccb1411605be

**Comment by jbrichau** [comment]
Addressed the 'undefined symbols' and 'messages sent but not implemented' in the Seaside-Gettext-Examples package by commenting that the WAGetTextExporter only works on Pharo. I left in the code but changed it so that the tests are no longer failing. Splitting the package would be too much for this issue.
